src/stream/src/executor/upstream_sink_union.rs

```diff
@@ -311,25 +422,79 @@ impl UpstreamSinkUnionExecutor {
                 "UpstreamSinkUnion",
             ]);
 
-        let upstreams = DynamicReceivers::new(inputs, Some(barrier_align), None);
+        let upstreams = DynamicReceivers::new(inputs, Some(barrier_align.clone()), None);
         pin_mut!(upstreams);
 
-        while let Some(msg) = upstreams.next().await {
-            yield msg?;
+        let mut current_barrier = None;
```

Comment:
`Option<Barrier>` is something like a queue with max capacity as 1. Can we instead use a unlimited capacity `VecDeque` to store all pending barriers so that we don't have to do conditional select on `self.barrier_rx.recv(), if current_barrier.is_none()`?